resources/RegexGrammar.pp

```diff
@@ -0,0 +1,207 @@
+//
+// Hoa
+//
+//
+// @license
+//
+// New BSD License
+//
+// Copyright © 2007-2017, Hoa community. All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//     * Redistributions of source code must retain the above copyright
+//       notice, this list of conditions and the following disclaimer.
+//     * Redistributions in binary form must reproduce the above copyright
+//       notice, this list of conditions and the following disclaimer in the
+//       documentation and/or other materials provided with the distribution.
+//     * Neither the name of the Hoa nor the names of its contributors may be
+//       used to endorse or promote products derived from this software without
+//       specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS AND CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+//
+// Grammar \Hoa\Regex\Grammar.
+//
+// Provide grammar of PCRE (Perl Compatible Regular Expression)for the LL(k)
+// parser. More informations at http://pcre.org/pcre.txt, sections pcrepattern &
+// pcresyntax.
+//
+// @copyright  Copyright © 2007-2017 Hoa community.
+// @license    New BSD License
+//
+
+
+// Skip.
+%skip   nl                       \n
+
+// Character classes.
+%token  negative_class_          \[\^               -> class
+%token  class_                   \[                 -> class
+%token class:posix_class         \[\^?:[a-z]+:\]
+%token class:class_              \[
+%token class:_class              \]                 -> default
+%token class:range               \-
+%token class:class_literals      [^\\\[\]-]+
+%token class:escaped_end_class   \\\]
+// taken over from literals
+%token class:character           \\([aefnrt]|c[\x00-\x7f])
+%token class:dynamic_character   \\([0-7]{3}|x[0-9a-zA-Z]{2}|x{[0-9a-zA-Z]+})
+%token class:character_type      \\([CdDhHNRsSvVwWX]|[pP]{[^}]+})
```

Comment:
@staabm I think something here messes up with your new numeric-string detection (see test failures). I'm not sure what though. If you have a chance to take a look I'd appreciate as that ast walking code is really kinda hell to look at for me as I have no clue about the AST walker and not much clue about phpstan internals :D

Reply:
I can have a look

Reply:
With the hoa.php trick above I think I can work it out..

Reply:
before this PR the ast for `'~^(02,([\d.]{10}))$~'` looked like:

```
$ php hoa.php
>  #expression
>  >  #concatenation
>  >  >  token(literal, ~)
>  >  >  token(anchor, ^)
>  >  >  #capturing
>  >  >  >  #concatenation
>  >  >  >  >  token(literal, 0)
>  >  >  >  >  token(literal, 2)
>  >  >  >  >  token(literal, ,)
>  >  >  >  >  #capturing
>  >  >  >  >  >  #quantification
>  >  >  >  >  >  >  #class
>  >  >  >  >  >  >  >  token(character_type, \d)
>  >  >  >  >  >  >  >  token(literal, .)
>  >  >  >  >  >  >  token(exactly_n, {10})
>  >  >  token(anchor, $)
```

after this PR it looks like

```
$ php hoa.php
>  #expression
>  >  #concatenation
>  >  >  token(literal, ~)
>  >  >  token(anchor, ^)
>  >  >  #capturing
>  >  >  >  #concatenation
>  >  >  >  >  token(literal, 0)
>  >  >  >  >  token(literal, 2)
>  >  >  >  >  token(literal, ,)
>  >  >  >  >  #capturing
>  >  >  >  >  >  #quantification
>  >  >  >  >  >  >  #class
>  >  >  >  >  >  >  >  token(class:character_type, \d)
>  >  >  >  >  >  >  >  token(class:class_literals, .)
>  >  >  >  >  >  >  token(exactly_n, {10})
>  >  >  token(anchor, $)
>  >  >  token(literal, ~)
```

the change in token-names need to be reflected in the `RegexArrayShapeMatcher`. I guess you just need to search for the hardcoded string values

Reply:
Ok that helped a LOT, thanks. It would have saved me a lot of guesswork earlier 😅  But anyway glad I got it now.